File: node/src/runner.js
const fs = require("fs");
const path = require("path");

const { cleanText } = require("./core");
const { loadRules } = require("./rules");
const metadata = require("./metadata");
const office = require("./office");
const { tooLarge, isSymlink, writeAtomic } = require("./safeio");

const utf8Decoder = new TextDecoder("utf-8", { fatal: true });

const PERMISSION_MESSAGE = "skipped (permission denied)";
const IO_MESSAGE = "skipped (i/o error)";

function byUtf8Bytes(a, b) {
  return Buffer.compare(Buffer.from(a.name, "utf-8"), Buffer.from(b.name, "utf-8"));
}

function walk(directory, exclude, acc, unreadable) {
  let entries;
  try {
    entries = fs.readdirSync(directory, { withFileTypes: true });
  } catch (error) {
    unreadable.push(directory);
    return;
  }
  entries.sort(byUtf8Bytes);
  for (const entry of entries) {
    if (exclude.includes(entry.name)) continue;
    const full = path.join(directory, entry.name);
    if (entry.isDirectory()) {
      walk(full, exclude, acc, unreadable);
    } else if (entry.isFile()) {
      acc.push(full);
    } else if (entry.isSymbolicLink()) {
      try {
        if (fs.statSync(full).isFile()) acc.push(full);
      } catch (error) {}
    }
  }
}

function collectFiles(paths, config) {
  const textExt = new Set(config.text_extensions.map((e) => e.toLowerCase()));
  const imageExt = new Set(config.image_extensions.map((e) => e.toLowerCase()));
  const documentExt = new Set((config.document_extensions || []).map((e) => e.toLowerCase()));
  const exclude = config.exclude || [];
  const all = [];
  const missing = [];
  const unreadable = [];
  for (const p of paths) {
    let stat;
    try {
      stat = fs.statSync(p);
    } catch (error) {
      if (error.code === "ENOENT" || error.code === "ENOTDIR") missing.push(p);
      else unreadable.push(p);
      continue;
    }
    if (stat.isFile()) all.push(p);
    else if (stat.isDirectory()) walk(p, exclude, all, unreadable);
    else missing.push(p);
  }
  const textFiles = [];
  const imageFiles = [];
  const documentFiles = [];
  const seen = new Set();
  for (const file of all) {
    let key;
    try {
      key = fs.realpathSync(file);
    } catch (error) {
      key = path.resolve(file);
    }
    if (seen.has(key)) continue;
    seen.add(key);
    const ext = path.extname(file).toLowerCase();
    if (textExt.has(ext)) textFiles.push(file);
    else if (imageExt.has(ext)) imageFiles.push(file);
    else if (documentExt.has(ext)) documentFiles.push(file);
  }
  return { textFiles, imageFiles, documentFiles, missing, unreadable };
}

function skipReport(file, message) {
  return {
    path: file,
    findings: [{ layer: "io", kind: "read", severity: "warn", message, count: 1, examples: [] }],
    changed: false,
    counts: { fixed: 0, warn: 1, error: 0 },
    has_errors: false,
  };
}

function safely(file, action) {
  try {
    return action();
  } catch (error) {
    if (error && (error.code === "EACCES" || error.code === "EPERM")) return skipReport(file, PERMISSION_MESSAGE);
    if (error && typeof error.code === "string") return skipReport(file, IO_MESSAGE);
    throw error;
  }
}

function processTextFile(file, config, rules, write) {
  if (tooLarge(file, config)) {
    return skipReport(file, "skipped (larger than max_file_bytes)");
  }
  const raw = fs.readFileSync(file);
  let original;
  try {
    original = utf8Decoder.decode(raw);
  } catch (error) {
    return skipReport(file, "skipped (not utf-8 text)");
  }
  const { text, report } = cleanText(original, config, rules, file);
  if (write && report.changed) {
    if (isSymlink(file)) {
      report.changed = false;
      report.findings.push({ layer: "io", kind: "write", severity: "warn", message: "refused to write through symlink", count: 1, examples: [] });
      report.counts.warn += 1;
      return report;
    }
    if (config.backup !== false) backup(file);
    writeAtomic(file, text);
  }
  return report;
}

function backup(file) {
  const bak = `${file}.bak`;
  if (!fs.existsSync(bak)) fs.copyFileSync(file, bak);
}

function run(paths, config, write) {
  const rules = loadRules();
  const { textFiles, imageFiles, documentFiles, missing, unreadable } = collectFiles(paths, config);
  const reports = [];
  for (const p of missing) reports.push(skipReport(p, "path not found"));
  for (const p of unreadable) reports.push(skipReport(p, PERMISSION_MESSAGE));
  for (const file of textFiles) reports.push(safely(file, () => processTextFile(file, config, rules, write)));
  const doBackup = config.backup !== false;
  const stripIcc = config.strip_icc === true;
  for (const file of imageFiles) {
    if (tooLarge(file, config)) {
      reports.push(skipReport(file, "skipped (larger than max_file_bytes)"));
      continue;
    }
    if (write) reports.push(safely(file, () => metadata.cleanFile(file, true, doBackup, stripIcc)));
    else reports.push(safely(file, () => metadata.inspectFile(file, stripIcc)));
  }
  for (const file of documentFiles) {
    if (tooLarge(file, config)) {
      reports.push(skipReport(file, "skipped (larger than max_file_bytes)"));
      continue;
    }
    if (write) reports.push(safely(file, () => office.cleanFile(file, true, doBackup)));
    else reports.push(safely(file, () => office.inspectFile(file)));
  }
  return reports;
}

module.exports = { run, collectFiles };

File: python/watermark_cleaner/runner.py
import os
from pathlib import Path

from .core import clean_text
from .findings import Finding, Report
from .layers import metadata, office
from .rules import load_rules
from .safeio import is_symlink, too_large, write_text_atomic

_PERMISSION_MESSAGE = "skipped (permission denied)"
_IO_MESSAGE = "skipped (i/o error)"


def collect_files(paths, config):
    text_ext = {e.lower() for e in config["text_extensions"]}
    image_ext = {e.lower() for e in config["image_extensions"]}
    document_ext = {e.lower() for e in config.get("document_extensions", [])}
    exclude = set(config.get("exclude", []))
    text_files = []
    image_files = []
    document_files = []
    missing = []
    unreadable = []
    for raw in paths:
        base = Path(raw)
        try:
            is_file = base.is_file()
            is_dir = base.is_dir()
        except OSError:
            unreadable.append(base)
            continue
        if is_file:
            _classify(base, text_ext, image_ext, document_ext, text_files, image_files, document_files)
        elif is_dir:
            for item in _walk(base, exclude, unreadable):
                _classify(item, text_ext, image_ext, document_ext, text_files, image_files, document_files)
        else:
            missing.append(base)
    return _dedupe(text_files), _dedupe(image_files), _dedupe(document_files), missing, unreadable


def _walk(directory, exclude, unreadable):
    try:
        with os.scandir(directory) as entries:
            children = sorted(entries, key=lambda entry: entry.name)
    except OSError:
        unreadable.append(directory)
        return
    for entry in children:
        if entry.name in exclude:
            continue
        path = Path(entry.path)
        try:
            if entry.is_dir(follow_symlinks=False):
                yield from _walk(path, exclude, unreadable)
            elif entry.is_file():
                yield path
        except OSError:
            unreadable.append(path)


def _dedupe(files):
    seen = set()
    result = []
    for item in files:
        try:
            key = item.resolve()
        except OSError:
            key = item.absolute()
        if key in seen:
            continue
        seen.add(key)
        result.append(item)
    return result


def _classify(item, text_ext, image_ext, document_ext, text_files, image_files, document_files):
    suffix = item.suffix.lower()
    if suffix in text_ext:
        text_files.append(item)
    elif suffix in image_ext:
        image_files.append(item)
    elif suffix in document_ext:
        document_files.append(item)


def _skip_report(path, message):
    report = Report(path=str(path))
    report.findings.append(Finding("io", "read", "warn", message, 1))
    return report


def _safely(path, action):
    try:
        return action()
    except PermissionError:
        return _skip_report(path, _PERMISSION_MESSAGE)
    except OSError:
        return _skip_report(path, _IO_MESSAGE)


def process_text_file(path, config, rules, write):
    if too_large(path, config):
        return _skip_report(path, "skipped (larger than max_file_bytes)")
    raw = Path(path).read_bytes()
    try:
        original = raw.decode("utf-8")
    except UnicodeDecodeError:
        return _skip_report(path, "skipped (not utf-8 text)")

    cleaned, report = clean_text(original, config=config, rules=rules, path=str(path))
    if write and report.changed:
        if is_symlink(path):
            report.changed = False
            report.findings.append(Finding("io", "write", "warn", "refused to write through symlink", 1))
            return report
        if config.get("backup", True):
            _backup(path)
        write_text_atomic(path, cleaned)
    return report


def run(paths, config, write):
    rules = load_rules()
    text_files, image_files, document_files, missing, unreadable = collect_files(paths, config)
    reports = []
    for path in missing:
        reports.append(_skip_report(path, "path not found"))
    for path in unreadable:
        reports.append(_skip_report(path, _PERMISSION_MESSAGE))
    for path in text_files:
        reports.append(_safely(path, lambda: process_text_file(path, config, rules, write)))
    backup = config.get("backup", True)
    strip_icc = config.get("strip_icc", False)
    for path in image_files:
        if too_large(path, config):
            reports.append(_skip_report(path, "skipped (larger than max_file_bytes)"))
            continue
        if write:
            reports.append(_safely(path, lambda: metadata.clean_file(path, write=True, backup=backup, strip_icc=strip_icc)))
        else:
            reports.append(_safely(path, lambda: metadata.inspect_file(path, strip_icc=strip_icc)))
    for path in document_files:
        if too_large(path, config):
            reports.append(_skip_report(path, "skipped (larger than max_file_bytes)"))
            continue
        if write:
            reports.append(_safely(path, lambda: office.clean_file(path, write=True, backup=backup)))
        else:
            reports.append(_safely(path, lambda: office.inspect_file(path)))
    return reports


def _backup(path):
    p = Path(path)
    backup = p.with_suffix(p.suffix + ".bak")
    if not backup.exists():
        backup.write_bytes(p.read_bytes())
